### PR TITLE
7903666: Jextract should report issues with missing dependencies

### PR DESCRIPTION
**Pull request by mcimadamore**
When custom include filters are specified on the command line, it is possible for jextract to generate non-sensical code (now even more so than in the past). Possible issues:

* a function descriptor refers to a non-existent struct/union
* a struct/union typedef refers to a non-existent struct/union
* a struct/union field refers to a non-existent struct/union
* a global variable refers to a non-existent struct/union

In such cases, jextract should terminate (and report the problem to the user), rather than generating buggy code.

The fix is relatively simple, and amounts at checking whether a declaration depends on a skipped struct in `IncludeHelper`. That said, when we detect an issue, we should stop jextract. So we need some way to notify jextract that some error occurred, and stop.

I initially had a hacky approach with some mutable field in `IncludeFilter`, but I've decided to implement a minimal Logger class, and make _all_ error/warnings go through the Logger. This way we can easily check how many errors occurred, and stop jextract if needed. Of course, this led to several changes, as I realized that we were being inconsistent on our usage of the property file. I now made sure that all messages use some key in the property file (e.g. no lone strings). This might be important if jextract message will ever be translated to languages other than English.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (no review required)

### Issue
 * [CODETOOLS-7903666](https://bugs.openjdk.org/browse/CODETOOLS-7903666): Jextract should report issues with missing dependencies (**Bug** - P3)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jextract.git pull/214/head:pull/214` \
`$ git checkout pull/214`

Update a local copy of the PR: \
`$ git checkout pull/214` \
`$ git pull https://git.openjdk.org/jextract.git pull/214/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 214`

View PR using the GUI difftool: \
`$ git pr show -t 214`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jextract/pull/214.diff">https://git.openjdk.org/jextract/pull/214.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jextract/pull/214#issuecomment-1948663270)

### CODE REVIEW

File: src/main/java/org/openjdk/jextract/impl/IncludeFilter.java [hunk 1]
```diff
@@ -93,11 +100,26 @@ public Void visitVariable(Declaration.Variable tree, Declaration parent) {
...
+        warnMissingDep(parent != null ? parent : tree, tree.type());
         return null;
     }
 
     @Override
     public Void visitDeclaration(Declaration decl, Declaration parent) {
         return null;
     }
+
+    void warnMissingDep(Declaration decl, Type type) {
```
Comment (mcimadamore) [inline src/main/java/org/openjdk/jextract/impl/IncludeFilter.java:112]: This is the meat of the fix

File: src/main/java/org/openjdk/jextract/impl/OutputFactory.java [hunk 2]
```diff
@@ -64,11 +64,7 @@ JavaSourceFile[] generate(Declaration.Scoped decl) {
     }
 
     private void generateDecl(Declaration tree) {
-        try {
-            tree.accept(this, null);
-        } catch (Exception ex) {
-            ex.printStackTrace();
-        }
+        tree.accept(this, null);
```
Comment (mcimadamore) [inline src/main/java/org/openjdk/jextract/impl/OutputFactory.java:67]: the try/catch seemed suspicious, and was also printing on std err, so I dropped that (after all, if some unexpected exception happens, we want to know?)

File: src/main/java/org/openjdk/jextract/JextractTool.java [hunk 3]
```diff
@@ -115,34 +103,34 @@ private static Path generateTmpSource(List<Path> headers) {
...
+                .map(new IncludeFilter(includeHelper, logger)::scan)
                 .map(new DuplicateFilter()::scan)
                 .map(new NameMangler(headerName)::scan)
-                .map(new UnsupportedFilter(errStream)::scan)
+                .map(new UnsupportedFilter(logger)::scan)
                 .findFirst().get();
-        return OutputFactory.generateWrapped(transformedDecl, targetPkg, libs, useSystemLoadLibrary);
+        return logger.hasErrors() ?
+                List.of() :
+                List.of(OutputFactory.generateWrapped(transformedDecl, targetPkg, libs, useSystemLoadLibrary));
```
Comment (JornVernee) [inline src/main/java/org/openjdk/jextract/JextractTool.java:133]: So, we now don't generate anything when there are errors.

I think we should also return a non-zero exit code in that case.

File: src/main/java/org/openjdk/jextract/impl/OutputFactory.java (hunk 2 above)
Comment (JornVernee) [reply src/main/java/org/openjdk/jextract/impl/OutputFactory.java:67]: I suppose this was added when jextract was less stable, or lacked filtering support. So, if we got an exception while generating one of the trees, jextract would still output _something_

File: test/testng/org/openjdk/jextract/test/toolprovider/includeDeps/TestBadIncludes.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        Path output = getOutputFilePath("TestBadIncludes-badIncludes.h");
+        Path outputH = getInputFilePath("bad_includes.h");
+        result = run(output,
+        "--include-struct", "B",
+                "--include-function", "m",
+                "--include-typedef", "T",
+                "--include-struct", "C",
+                "--include-function", "n",
+                outputH.toString());
+    }
```
Comment (JornVernee) [inline test/testng/org/openjdk/jextract/test/toolprovider/includeDeps/TestBadIncludes.java:47]: Maybe `result.checkFailure()` here too?

**Review by JornVernee — APPROVED** [review]

File: test/testng/org/openjdk/jextract/test/toolprovider/includeDeps/TestBadIncludes.java (hunk 4 above)
Comment (mcimadamore) [reply test/testng/org/openjdk/jextract/test/toolprovider/includeDeps/TestBadIncludes.java:47]: whoops - yes!

File: src/main/java/org/openjdk/jextract/JextractTool.java (hunk 3 above)
Comment (mcimadamore) [reply src/main/java/org/openjdk/jextract/JextractTool.java:133]: True

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**